### PR TITLE
[KeyVault] adding recovery feature commands

### PR DESCRIPTION
**Pull request by schaabs**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py [hunk 1]
```diff
@@ -237,9 +240,44 @@ def create_keyvault(client,
...
+                                      StoragePermissions.get,
+                                      StoragePermissions.list,
+                                      StoragePermissions.delete,
+                                      StoragePermissions.set,
+                                      StoragePermissions.update,
+                                      StoragePermissions.regeneratekey,
+                                      StoragePermissions.setsas,
+                                      StoragePermissions.listsas,
+                                      StoragePermissions.getsas,
+                                      StoragePermissions.deletesas])
```
Comment (troydai) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py]: Is there a way to get these options from SDK?

File: src/command_modules/azure-cli-role/setup.py [hunk 2]
```diff
@@ -34,7 +34,7 @@
     'azure-cli-core',
     'azure-mgmt-authorization==0.30.0rc6',
     'azure-graphrbac==0.30.0rc6',
-    'azure-keyvault==0.3.0',
+    'azure-keyvault==0.3.4',
```
Comment (troydai) [inline src/command_modules/azure-cli-role/setup.py]: @derekbekoe raise any API profile issue?

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py (hunk 1 above)
Comment (schaabs) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py]: All the permissions enumerations used here are defined in the SDK, we explicitly list out the values because the permissions granted by default are not all available permissions.

File: src/command_modules/azure-cli-role/setup.py (hunk 2 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/setup.py]: Changing the API versions of other command modules may break their tests depending on whether there were breaking changes between the different SDK versions.

Comment (schaabs) [inline src/command_modules/azure-cli-role/setup.py]: there are no breaking changes between azure-keyvault 0.3.0 and 0.3.4

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py [hunk 3]
```diff
@@ -137,6 +137,9 @@ def register_attributes_argument(scope, name, attr_class, create=False, ignore=N
 register_cli_argument('keyvault', 'enabled_for_template_deployment',
                       help='Allow Resource Manager to retrieve secrets from the vault.',
                       **three_state_flag())
+register_cli_argument('keyvault', 'enable_for_soft_delete',
+                      help='Enable vault deletion recovery for the vault, and all contained entities',
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py]: I'd recommend shortening the option name for this to "--enable-soft-delete".

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py [hunk 4]
```diff
@@ -164,6 +167,11 @@ def register_attributes_argument(scope, name, attr_class, create=False, ignore=N
                       help='Space separated list. Possible values: {}'.format(
                           certificate_permission_values), arg_group='Permission')
 
+# purge arguments
+register_cli_argument('keyvault purge', 'vault_name', required=True, completer=None, validator=None)
+register_cli_argument('keyvault purge', 'location', required=True, completer=None, validator=None)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py]: What is the reason for these two registrations?

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/commands.py [hunk 5]
```diff
@@ -44,13 +47,19 @@
                                 data_client_path.format('KeyVaultClient.get_keys'))
 cli_keyvault_data_plane_command('keyvault key list-versions',
                                 data_client_path.format('KeyVaultClient.get_key_versions'))
+cli_keyvault_data_plane_command('keyvault key list-deleted',
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/commands.py]: Alternatively, you could simply enhance the existing list command with a --deleted flag (or something similar) rather than have a completely separate command. This would be similar to list/list-all folding that many commands do

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py [hunk 6]
```diff
@@ -501,6 +540,24 @@ def download_secret(client, vault_base_url, secret_name, file_path, encoding=Non
...
+def backup_secret(client, vault_base_url, secret_name, file_path):
+    backup = client.backup_secret(vault_base_url, secret_name).value
+    with open(file_path, 'wb') as output:
+        output.write(backup)
+
+
+backup_secret.__doc__ = KeyVaultClient.backup_secret.__doc__
+
+
+def restore_secret(client, vault_base_url, file_path):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py]: You should add a scenario tests that backs up and restores a secret successfully. From doing the original KV commands, I would be surprised if more edge cases didn't need to be covered on account of Python2 vs 3 and Linux vs Windows.

**Comment by schaabs** [comment]
@tjprescott All the commands have tests, and I've made updates based on your comments.  PTAL

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py (hunk 6 above)
Comment (schaabs) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py]: Backup and restore has been added to the secret scenaro

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/commands.py (hunk 5 above)
Comment (schaabs) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/commands.py]: list and list-deleted have some conflicting arguments in some cases and return different types in all, this is why I created them as separate commands

File: src/command_modules/azure-cli-keyvault/HISTORY.rst [hunk 7]
```diff
@@ -2,13 +2,23 @@
 
 Release History
 ===============
+2.0.7 (2017-06-16)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/HISTORY.rst]: This should just say "unreleased". Derek will fill this in later.

File: src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py [hunk 8]
```diff
@@ -209,11 +210,37 @@ def get_default_policy(client, scaffold=False):  # pylint: disable=unused-argume
...
-def create_keyvault(client,
+def recover_keyvault(client, vault_name, resource_group_name, location):
+    """
+    Recover a previously deleted keyvault for which soft delete was enabled
+    :param client:
+    :param vault_name: The name of the vault to recover
+    :param resource_group_name: The original resource group of the vault to recover
+    :param location: The original location of the vault to recover
+    :return: The properties of the recovered key vault
+    """
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py]: This help text should be moved (and reformatted) to the help.py file for the module. This is because the live docs links to help.py if someone wants to make a change to help text.  While this way of creating help text is still supported (for SDK reflection) we don't allow it in custom commands for new PRs.

File: src/command_modules/azure-cli-keyvault/tests/test_keyvault_commands.py [hunk 9]
```diff
@@ -211,22 +212,6 @@ def body(self):
...
-        self.cmd(
-            'keyvault key import --vault-name {} -n import-key-plain --pem-file "{}" -p software'.format(
-                kv, key_plain_file))
-        self.cmd(
-            'keyvault key import --vault-name {} -n import-key-encrypted --pem-file "{}" --pem-password {} -p hsm'.format(
-                kv, key_enc_file, key_enc_password))
-
-        byok_key_file = os.path.join(TEST_DIR, 'TestBYOK-NA.byok')
-        self.cmd(
-            'keyvault key import --vault-name {} -n import-key-byok --byok-file "{}"'.format(kv,
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/tests/test_keyvault_commands.py]: I understand why you removed the BYOK test, but why the PEM file import?

File: src/command_modules/azure-cli-keyvault/tests/test_keyvault_commands.py [hunk 10]
```diff
@@ -587,5 +586,92 @@ def body(self):
                 kv, pfx_plain_file, pfx_policy_path))
 
 
+class KeyVaultSoftDeleteScenarioTest(ResourceGroupVCRTestBase):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-keyvault/tests/test_keyvault_commands.py]: Ideally this would inherit from ScenarioTest, but I'm fine merging this as is.
